### ISSUE TITLE
Exporting ONNX model with captum heatmaps generation

### ISSUE THREAD

**Issue by Rayndell** [pytorch/pytorch#173369]
### 🐛 Describe the bug

I have trouble exporting and loading an ONNX model which generate captum heatmaps as outputs. My export code looks like this:

```python
class ClassificationModel(torch.nn.Module):
    def __init__(self, base_model, device, resize_shape=None, heatmap_generator=None):
        super(ClassificationModel, self).__init__()
        self.base_model = base_model
        self.device = device
        self.transform = get_preprocessing(resize_shape=resize_shape, data_augmentation=False)
        self.heatmap_generator = heatmap_generator
        
    def forward(self, inputs):
        # Apply transforms
        inputs = self.transform(inputs)
        
        # Apply model to transformed input
        outputs = self.base_model(inputs)
        
        # Generate heatmaps
        heatmaps = torch.zeros(inputs.shape)
        if self.heatmap_generator:
            heatmaps = self.heatmap_generator.generate(inputs, outputs.shape[1], self.device)
            # Resize heatmap if needed
            img_height, img_width = inputs.shape[-2:]
            resize = get_resize((img_height, img_width))
            heatmaps = resize(heatmaps)
        
        return outputs, heatmaps
    
    def export_to_onnx(self, path):
        self.eval()
        input_names = ["input"]
        output_names = ["output"]
        
        dynamic_axes = {
            "input": {0: "batch_size", 2: "height", 3: "width"},
        }
        
        if self.heatmap_generator:
            output_names.append("heatmaps")
            dynamic_axes["heatmaps"] = {0: "batch_size", 2: "height", 3: "width"}
           
        self.to(self.device)
        input_example = torch.randint(0, 255, (1, 3, 512, 512), dtype=torch.uint8).to(self.device)
        torch.onnx.export(self, input_example, path, export_params=True, opset_version=16,
            input_names=input_names, output_names=output_names,
            dynamic_axes=dynamic_axes, dynamo=False
        )

# Reading the configuration file
config = get_config(args.config_file)
resize_shape = (config.resize_height, config.resize_width) if config.resize else None
    
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
# Load the given checkpoint and retrieve the class names
checkpoint = torch.load(args.checkpoint_path, weights_only=True)
class_names = ["CD","CV","Pachy","SD","Hex","Max","Min"]
if "class_names" in checkpoint:
    class_names = checkpoint['class_names']
    
# Rebuild the model
model = build_model(config, len(class_names))
        
# Load the model weights
if "model_state_dict" in checkpoint:
    model.load_state_dict(checkpoint['model_state_dict'], strict=False)
else:
    model.load_state_dict(checkpoint, strict=False)
    
# Create a heatmap generator
heatmap_generator = HeatmapGenerator(model, heatmap_type="IntegratedGradients")
    
export_to_onnx(model, args.output, device, resize_shape, heatmap_generator)
```

The heatmaps can be successfully retrieved from the exported ONNX model during inference. But they are always the same whatever input image is provided. From what I understood there might be some parameters converted to constants during the export, so that the resulting heatmaps are constant and inferred from the input_example provided to the torch.onnx.export function. Here is the output during export:

C:\Users\aurelien.ADCIS-Caen\AppData\Local\Programs\Python\Python312\Lib\site-packages\captum\_utils\gradient.py:133: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert outputs[0].numel() == 1, (
C:\Users\aurelien.ADCIS-Caen\AppData\Local\Programs\Python\Python312\Lib\site-packages\captum\attr\_core\integrated_gradients.py:379: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  * torch.tensor(step_sizes).float().view(n_steps, 1).to(grad.device)

Does anyone have experienced the same issue, or have tried to export captum functions into ONNX. What can be a workaround to this?

Thank you very much.

### Versions

PyTorch version: 2.9.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Professionnel (10.0.26100 64 bits)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 581.80
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\bin\cudnn_ops64_9.dll
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
Manufacturer: GenuineIntel
Family: 198
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3696
MaxClockSpeed: 3696
L2CacheSize: 1536
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.17.0
[pip3] onnx-ir==0.1.12
[pip3] onnxruntime==1.23.2
[pip3] onnxruntime-gpu==1.23.2
[pip3] onnxscript==0.5.6
[pip3] optree==0.16.0
[pip3] segmentation_models_pytorch==0.5.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.9.1+cu126
[pip3] torchaudio==2.9.1+cu126
[pip3] torchmetrics==1.8.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.24.1+cu126
[conda] Could not collect

cc @justinchuby @titaiwangms @chauhang @penguinwu @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by justinchuby** [comment]
cc @angelayi
